geps/gep-2162.md

```diff
@@ -0,0 +1,114 @@
+# GEP-2162: Supported features in GatewayClass Status
+
+* Issue: [#2162](https://github.com/kubernetes-sigs/gateway-api/issues/2162)
+* Status: Provisional
+
+## TLDR
+
+This GEP proposes to enhance the [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185) to include a list of gw-api features supported by the installed Gateway Class. 
+
+## Goals
+
+* Enable users easily see what features the implementation(gateway class) support
+* Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+## Non-Goals
+
+* Validate correctness of supported features published by the implementation.
+
+## Introduction
+
+The current [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185 is only used to store conditions the controller publishes.
+
+Partnered with the [Conformance Profiles](https://github.com/kubernetes-sigs/gateway-api/blob/main/geps/gep-1709.md) work, we want to:
+
+1. Enable users to easily see what features an implementation supports
+1. Improve UX and potentially provide clear errors when a user is trying to use an unsupported feature
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **should** publish the supported/unsupported features before Accepting the GatewayClass.
+
+They can either maintain static lists of supported/unsupported features OR generate them from their conformance tests results/reports as part of their CI.
+
+## API
+
+This GEP proposes API changes describes as follow:
+1. Adding `FeaturesSupport` and  `GatewayClassFeatures` structs. 
+2. Update the `GatewayClassStatus` struct to include `GatewayClassFeatures`
+
+```go
+// FeaturesSupport contains the supported features and unsupported features by the installed GatewayClass.
+type FeaturesSupport struct {
+	// SupportedFeatures indicates which extended features were flagged as
+	// supported by the implementation and tests will be attempted for.
+	SupportedFeatures []string `json:"supportedFeatures,omitempty"`
```

Comment:
```suggestion
	// SupportedFeatures indicates which extended and experimental features
	// are supported by this GatewayClass. When running conformance tests,
	// tests that include features not listed here will be skipped.
	SupportedFeatures []string `json:"supportedFeatures,omitempty"`
```

Reply:
Actually on further thought, maybe this is even broader than "extended and experimental"... @shaneutt and @mlavacca may have some thoughts here. I'm not really sure how we want to handle the broader categories/profiles like mesh implementations that don't support Gateways. Maybe we should just say "which features are supported by this GatewayClass"?

Reply:
I copy pasted the description, my bad. 

Essentially the way I envision it as I wrote the`GatewayClassFeatures` struct.

`SupportedFeatures` is just a list of strings but a higher is holding it. See `Core` and `Extended` under the `GatewayClassFeatures`